In [1]:
#Initial imports 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler
from sklearn.metrics import accuracy_score, f1_score, recall_score, roc_auc_score, plot_confusion_matrix
from sklearn.metrics import confusion_matrix, plot_roc_curve, precision_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.decomposition import PCA

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier, BaggingClassifier, RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier, StackingRegressor, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB
import xgboost as xgb

pd.set_option('display.max_columns', None)

In [53]:
df = pd.read_sas('../../Data/nsch_2020_topical.sas7bdat')
mask = {
    1.0: 1.0,
    2.0: 0.0
}
df['Target'] = df['K2Q35A'].map(mask)
df.dropna(subset=['Target'], inplace=True)

y = df['Target']
X = df.drop(columns=['Target', 'K2Q35A', 'K2Q35A_1_YEARS', 'K2Q35B', 'K2Q35C', 'K2Q35D', 'AUTISMMED', 'AUTISMTREAT'])

drop_cols = []
for col in X.columns:
    null_count = X[col].isna().sum()
    null_perc = X[col].isna().sum() / len(X)
    if null_perc >= 0.5:
        drop_cols.append(col)

X = X.drop(columns=drop_cols)

In [57]:
drop_cols

['K6Q41R_STILL',
 'K6Q42R_NEVER',
 'K6Q43R_NEVER',
 'K6Q13A',
 'K6Q13B',
 'K6Q14A',
 'K6Q14B',
 'K4Q28X01',
 'K4Q28X02',
 'K4Q28X03',
 'K4Q28X_EAR',
 'K4Q28X04',
 'K4Q28X05',
 'SESPLANYR',
 'SESPLANMO',
 'K4Q37',
 'SPCSERVMO',
 'LIVEUSA_YR',
 'LIVEUSA_MO',
 'A1_LIVEUSA',
 'A2_LIVEUSA',
 'HANDS',
 'COORDINATION',
 'ERRANDALONE',
 'ALLERGIES_CURR',
 'ARTHRITIS_CURR',
 'K2Q40B',
 'K2Q61B',
 'K2Q41B',
 'K2Q42B',
 'HEART_BORN',
 'HEART_CURR',
 'HEADACHE_CURR',
 'K2Q38B',
 'K2Q33B',
 'K2Q32B',
 'SICKLECELL',
 'THALASSEMIA',
 'BLOOD_OTHER',
 'BLOOD_SCREEN',
 'CYSTFIB_SCREEN',
 'GENETIC_SCREEN',
 'K2Q34B',
 'K2Q36B',
 'K2Q60B',
 'K2Q37B',
 'K2Q30B',
 'K2Q31B',
 'K2Q31D',
 'ADDTREAT',
 'SEEKCARE',
 'CONFIRMINJURY',
 'K6Q40',
 'DOCPRIVATE',
 'K6Q10',
 'K6Q12',
 'NOTELIG',
 'AVAILABLE',
 'APPOINTMENT',
 'TRANSPORTCC',
 'NOTOPEN',
 'ISSUECOST',
 'SESCURRSVC',
 'K4Q38',
 'K5Q21',
 'TREATCHILD',
 'TREATADULT',
 'MEDHISTORY',
 'WRITEPLAN',
 'RECEIVECOPY',
 'HEALTHKNOW',
 'KEEPINSADULT',
 'K12Q01_A',


In [58]:
y.value_counts()

0.0    41343
1.0     1268
Name: Target, dtype: int64

1268 autistic children is likely enough to undersample instead of oversampling; which is ideal since we have many binary columns in our dataframe.

In [59]:
X.shape

(42611, 259)

In [61]:
for col in df.columns:
    null_count = df[col].isna().sum()
    null_perc = df[col].isna().sum() / len(df)
    print(f'{col} null count: {null_count}')
    print(f'{col} null percent: {null_perc}\n')

FIPSST null count: 0
FIPSST null percent: 0.0

STRATUM null count: 0
STRATUM null percent: 0.0

HHID null count: 0
HHID null percent: 0.0

FORMTYPE null count: 0
FORMTYPE null percent: 0.0

TOTKIDS_R null count: 0
TOTKIDS_R null percent: 0.0

TENURE null count: 0
TENURE null percent: 0.0

HHLANGUAGE null count: 151
HHLANGUAGE null percent: 0.0035436859027011803

SC_AGE_YEARS null count: 0
SC_AGE_YEARS null percent: 0.0

SC_SEX null count: 0
SC_SEX null percent: 0.0

K2Q35A_1_YEARS null count: 41453
K2Q35A_1_YEARS null percent: 0.9728239187064373

BIRTH_MO null count: 260
BIRTH_MO null percent: 0.0061017108258430925

BIRTH_YR null count: 379
BIRTH_YR null percent: 0.008894416934594354

MOMAGE null count: 907
MOMAGE null percent: 0.02128558353476802

K6Q41R_STILL null count: 32554
K6Q41R_STILL null percent: 0.763981131632677

K6Q42R_NEVER null count: 30586
K6Q42R_NEVER null percent: 0.717795874304757

K6Q43R_NEVER null count: 30633
K6Q43R_NEVER null percent: 0.718898875877121

K6Q13A nul

ONEWORD null count: 32064
ONEWORD null percent: 0.7524817535378189

TWOWORDS null count: 32064
TWOWORDS null percent: 0.7524817535378189

THREEWORDS null count: 32064
THREEWORDS null percent: 0.7524817535378189

ASKQUESTION null count: 32064
ASKQUESTION null percent: 0.7524817535378189

ASKQUESTION2 null count: 32064
ASKQUESTION2 null percent: 0.7524817535378189

TELLSTORY null count: 32064
TELLSTORY null percent: 0.7524817535378189

UNDERSTAND null count: 32064
UNDERSTAND null percent: 0.7524817535378189

DIRECTIONS null count: 32064
DIRECTIONS null percent: 0.7524817535378189

POINT null count: 32064
POINT null percent: 0.7524817535378189

DIRECTIONS2 null count: 32064
DIRECTIONS2 null percent: 0.7524817535378189

UNDERSTAND2 null count: 32064
UNDERSTAND2 null percent: 0.7524817535378189

RHYMEWORD null count: 35998
RHYMEWORD null percent: 0.8448053319565371

REPEATED null count: 12538
REPEATED null percent: 0.29424327051700266

K7Q30 null count: 12587
K7Q30 null percent: 0.295393208

K4Q02_R null percent: 0.1928140620966417

DENTISTVISIT null count: 8594
DENTISTVISIT null percent: 0.20168501091267513

K4Q22_R null count: 175
K4Q22_R null percent: 0.00410692074816362

TREATNEED null count: 36947
TREATNEED null percent: 0.8670765764708643

K4Q24_R null count: 277
K4Q24_R null percent: 0.006500668841378987

K4Q26 null count: 35679
K4Q26 null percent: 0.8373190021355988

C4Q04 null count: 199
C4Q04 null percent: 0.004670155593626059

HOSPITALER null count: 146
HOSPITALER null percent: 0.0034263453098965055

K4Q04_R null count: 204
K4Q04_R null percent: 0.004787496186430734

K5Q11 null count: 35236
K5Q11 null percent: 0.8269226256131046

K5Q20_R null count: 7044
K5Q20_R null percent: 0.16530942714322594

K5Q22 null count: 40884
K5Q22 null percent: 0.9594705592452653

K5Q30 null count: 22613
K5Q30 null percent: 0.5306845650184224

K5Q32 null count: 38983
K5Q32 null percent: 0.914857665860928

K5Q31_R null count: 6962
K5Q31_R null percent: 0.16338504142122925

K8Q21 null 

In [42]:
drop_cols

['K6Q41R_STILL',
 'K6Q42R_NEVER',
 'K6Q43R_NEVER',
 'K6Q13A',
 'K6Q13B',
 'K6Q14A',
 'K6Q14B',
 'K4Q28X01',
 'K4Q28X02',
 'K4Q28X03',
 'K4Q28X_EAR',
 'K4Q28X04',
 'K4Q28X05',
 'SESPLANYR',
 'SESPLANMO',
 'K4Q37',
 'SPCSERVMO',
 'LIVEUSA_YR',
 'LIVEUSA_MO',
 'A1_LIVEUSA',
 'A2_LIVEUSA',
 'HANDS',
 'COORDINATION',
 'ALLERGIES_CURR',
 'ARTHRITIS_CURR',
 'K2Q40B',
 'K2Q61B',
 'K2Q41B',
 'K2Q42B',
 'HEART_BORN',
 'HEART_CURR',
 'HEADACHE_CURR',
 'K2Q38B',
 'K2Q33B',
 'K2Q32B',
 'SICKLECELL',
 'THALASSEMIA',
 'BLOOD_OTHER',
 'BLOOD_SCREEN',
 'CYSTFIB_SCREEN',
 'GENETIC_SCREEN',
 'K2Q34B',
 'K2Q36B',
 'K2Q60B',
 'K2Q37B',
 'K2Q30B',
 'K2Q31B',
 'K2Q31D',
 'ADDTREAT',
 'SEEKCARE',
 'CONFIRMINJURY',
 'K6Q40',
 'DOCPRIVATE',
 'K6Q10',
 'K6Q12',
 'NOTELIG',
 'AVAILABLE',
 'APPOINTMENT',
 'TRANSPORTCC',
 'NOTOPEN',
 'ISSUECOST',
 'SESCURRSVC',
 'K4Q38',
 'TREATADULT',
 'RECEIVECOPY',
 'KEEPINSADULT',
 'K12Q01_A',
 'K12Q01_B',
 'K12Q01_C',
 'K12Q01_D',
 'K12Q01_E',
 'K12Q01_F',
 'K12Q01_G',
 'ONEWO

In [62]:
X.head()

FIPSST STRATUM         HHID FORMTYPE  TOTKIDS_R  TENURE  HHLANGUAGE  \
0  b'17'    b'1'  b'20000003'    b'T1'        2.0     1.0         1.0   
1  b'29'   b'2A'  b'20000004'    b'T3'        1.0     1.0         1.0   
2  b'47'    b'1'  b'20000005'    b'T1'        1.0     1.0         1.0   
3  b'28'    b'1'  b'20000014'    b'T3'        2.0     1.0         1.0   
4  b'55'    b'1'  b'20000015'    b'T3'        2.0     2.0         1.0   

   SC_AGE_YEARS  SC_SEX  BIRTH_MO  BIRTH_YR  MOMAGE  K4Q32X01  K4Q32X02  \
0           3.0     1.0       8.0    2017.0    26.0       NaN       NaN   
1          14.0     2.0       2.0    2006.0    31.0       NaN       NaN   
2           1.0     2.0      10.0    2018.0    28.0       NaN       NaN   
3          15.0     2.0      10.0    2004.0    29.0       1.0       2.0   
4          16.0     2.0       8.0    2004.0    24.0       1.0       2.0   

   K4Q32X03  K4Q32X04  K4Q32X05  DENTALSERV1  DENTALSERV2  DENTALSERV3  \
0       NaN       NaN       NaN          NaN          NaN          NaN   
1       NaN       NaN       NaN          1.0          1.0          1.0   
2       NaN       NaN       NaN          NaN          NaN          NaN   
3       2.0       2.0       2.0          1.0          1.0          2.0   
4       2.0       2.0       2.0          NaN          NaN          NaN   

   DENTALSERV4  DENTALSERV5  DENTALSERV6  DENTALSERV7  K11Q43R  A1_AGE  \
0          NaN          NaN          NaN          NaN      1.0    29.0   
1          1.0          1.0          2.0          2.0      1.0    43.0   
2          NaN          NaN          NaN          NaN      0.0    30.0   
3          1.0          2.0          2.0          2.0      1.0    44.0   
4          NaN          NaN          NaN          NaN      2.0    42.0   

   A2_AGE  HHCOUNT  FAMCOUNT  BREATHING  SWALLOWING  STOMACH  PHYSICALPAIN  \
0    33.0      4.0       4.0        2.0         2.0      2.0           2.0   
1     NaN      3.0       2.0        2.0         2.0      2.0           2.0   
2    43.0      3.0       3.0        2.0         2.0      2.0           2.0   
3    43.0      4.0       4.0        2.0         2.0      1.0           2.0   
4    42.0      4.0       4.0        2.0         2.0      NaN           2.0   

   TOOTHACHES  GUMBLEED  CAVITIES  MEMORYCOND  WALKSTAIRS  DRESSING  K2Q43B  \
0         2.0       2.0       2.0         NaN         NaN       NaN     2.0   
1         2.0       2.0       2.0         2.0         2.0       2.0     2.0   
2         2.0       2.0       2.0         NaN         NaN       NaN     2.0   
3         1.0       2.0       1.0         2.0         2.0       2.0     2.0   
4         2.0       2.0       2.0         2.0         2.0       2.0     2.0   

   BLINDNESS  ALLERGIES  ARTHRITIS  K2Q40A  K2Q61A  K2Q41A  K2Q42A  HEART  \
0        2.0        2.0        2.0     2.0     2.0     2.0     2.0    2.0   
1        2.0        2.0        2.0     1.0     2.0     2.0     2.0    2.0   
2        2.0        2.0        2.0     2.0     2.0     2.0     2.0    2.0   
3        2.0        2.0        2.0     2.0     2.0     2.0     2.0    2.0   
4        2.0        2.0        2.0     2.0     2.0     2.0     2.0    2.0   

   HEADACHE  K2Q38A  K2Q33A  K2Q32A  DOWNSYN  BLOOD  CYSTFIB  GENETIC  K2Q34A  \
0       2.0     2.0     2.0     2.0      2.0    2.0      2.0      2.0     2.0   
1       2.0     2.0     1.0     2.0      2.0    2.0      2.0      2.0     2.0   
2       2.0     2.0     2.0     2.0      2.0    2.0      2.0      2.0     2.0   
3       2.0     2.0     2.0     2.0      2.0    2.0      2.0      2.0     2.0   
4       2.0     2.0     1.0     2.0      2.0    2.0      2.0      2.0     2.0   

   K2Q36A  K2Q60A  K2Q37A  K2Q30A  K2Q31A  CONCUSSION  K2Q05  S4Q01  \
0     2.0     2.0     2.0     2.0     2.0         2.0    1.0    1.0   
1     2.0     2.0     2.0     2.0     2.0         2.0    2.0    1.0   
2     2.0     2.0     2.0     2.0     2.0         2.0    2.0    1.0   
3     2.0     2.0     2.0     2.0     2.0   

In [66]:
X[0:10].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10 entries, 0 to 9
Columns: 259 entries, FIPSST to FWC
dtypes: float64(255), object(4)
memory usage: 20.3+ KB
